libpod/container_api.go

```diff
@@ -639,6 +639,30 @@ func (c *Container) Cleanup() error {
 	return c.cleanup()
 }
 
+// Delete asks the OCI runtime to release this container resources
+// It also runs the post-stop hooks
+func (c *Container) Delete(ctx context.Context) error {
+       if !c.batched {
+               c.lock.Lock()
+               defer c.lock.Unlock()
+               if err := c.syncContainer(); err != nil {
+                       return err
+               }
+       }
+
+       // Check if state is good
+       if c.state.State == ContainerStateRunning || c.state.State == ContainerStatePaused {
+               return errors.Wrapf(ErrCtrStateInvalid, "container %s is running or paused, refusing to delete", c.ID())
+       }
+
+       // Check if we have active exec sessions
+       if len(c.state.ExecSessions) != 0 {
+               return errors.Wrapf(ErrCtrStateInvalid, "container %s has active exec sessions, refusing to delete", c.ID())
+       }
+
+       return c.delete(ctx)
```

Comment:
I think we need to include the missing parts from https://github.com/containers/libpod/blob/master/libpod/container_internal.go#L546-L564 here to make sure we leave the state in a good place. `removeConmonFiles()` and the state reset code/`c.save()` make sure we can restart the container after it stops.

Might want to refactor that into a function so both reinit() and Delete() can call it.

Reply:
Ack. But which path are we going to follow? Refactor and keep the new Delete public method or extend cleanup() (or Cleanup()?) to eventually call delete?

Reply:
Actually, the reset code/c.save() make `podman ps` return `Created` in the `STATUS` column instead of `Exited ($EXIT_CODE)...`, which is not what I would expect from a container that has run at least once.

The `removeConmonFiles` part seems fine though.

Reply:
Hm. We do need the state reset... Maybe we need a ContainerStateDeleted to properly accomodate this. I'll think about this for a bit during lunch.